### DISCUSSION TITLE
Proposal: remove distinctness requirement from `multi` properties

### DISCUSSION THREAD

**Discussion by elprans** [geldata/gel#2054] [Pre-RFC]
**Update:** implementation PR https://github.com/edgedb/edgedb/pull/2853

-------

Background
----------

All values in EdgeQL are considered to be multisets, which are a generalization of the set concept that allows duplicate elements.  For brevity we refer to multisets as just "sets", and use "proper set" or "distinct set" when referring to proper mathematical sets.

In EdgeDB the data is modeled as a directed graph, where objects are vertices and links are edges.  Query path expressions (`Foo.bar.baz`) are graph traversal operators and, by definition, always return a *distinct set*.  This means that there can be no link of the same name between a pair of objects, and this also means that a set of objects pointed to by a `multi` link is always distinct.  Computed links must follow this rule too so as to behave exactly like materialized links externally.  To illustrate, consider a common case of finding a set of "friends of friends" of a particular user.  In EdgeQL this is simply `SELECT User.friends.friends`.  Proper set semantics here allows this operation to use a more efficient semi-join in place of a regular join (especially if the latter is later coupled with a use of `DISTINCT`).  A majority of queries on object paths are like that and so this behavior is desirable.

Properties are different, because their value is scalar, and scalars lack true identity, which makes any proper set mechanics on scalar sets expensive due to explicit elimination of duplicates.  Furthermore, non-distinct scalar sets are actually *desirable* in many queries, especially where analytics and tuples are involved.  Finally, we should enforce consistent multiplicity rules on computed properties, and static inference of multiplicity on scalars is weak and will effectively force users to pollute queries with pointless and expensive `DISTINCT`. 

Proposal
--------

Allow multiplicity greater than one in `multi` properties.

As outlined above, non-distinct `multi` properties are useful in queries, and so from the rule of equivalence of computed and material relations, stored `multi` properties should be allowed to be non-distinct as well.

Impact
------

Stored `multi` properties will no longer complain about duplicates in input.  Computed properties currently lack multiplicity validation anyway, so there is no change in behavior there.

~The `-=` in-place `UPDATE` operation is no longer available for `multi` properties due to comprehension and implementation challenges.  Retaining the original equivalence of `<prop> -= <arg>` to `(SELECT <prop> FILTER <prop> != <arg>)` is confusing because it is a proper set operation on a multiset.  Changing it to mean `<prop>.pop(foo) for foo in <arg>` is hard to implement efficiently and might arguably also not be universally intuitive.~

Digression: why multi properties are needed at all
-------------------------------------------------

Here are some data points about `multi` propertes contrasted with arrays and links to objects:

`multi` property *pros* vs arrays:

1. `multi` properties allow efficient search and mutation of large sets.  Arrays are much slower for those operations.
2. `multi` properties can have indexes and constraints appied to individual elements; arrays, in general, cannot.

`multi` property *cons* vs arrays:

1. On small sets, arrays are faster to retrieve.
2. It's easier to retain the original order in arrays.

`multi` propery *pros* vs objects:

1. Objects require extra 32 bytes (`id` + `__type__`) per entry.
2. Forward `multi` links would require an extra `join` (link table + object table) in contrast to only one join for `multi` properties (this is _significant_).
3. Emulation of multi links via a backward single link to "owner" object requires extra boilerplate to define the computable in the parent object and longer more awkward path expressions in queries.

`multi` property *cons* vs objects:

1. If there is a lot of duplication in property values, using objects with an `exclusive` constraint on the value would be more space-efficient.
2. If you end up needing to assocociate more than one value with each, object proxies are much easier to migrate.

Bottom line: _stored_ `multi` properties are the most efficient way to represent certain types of data.

Distinct opt-in
---------------

With all the above taken into account, there are cases where distinct multi properties are useful.  With the proposed change implemented, one way to opt back into guaranteeing `multi` property set distinctness is to declare an explicit `constraint exclusive on (@source, @target)`. However, this is not entirely equivalent to the current distinct behavior, as an attempt to add a duplicate value to a set would result in `ConstraintViolationError` instead of being a no-op.  Thus, we also consider adding a declaration modifier (`distinct`) to properties to signal an opt-in into current "proper set" behavior: `distinct multi property foo -> str`.

**Comment by vpetrovykh** [comment]
A few minor comments:

> Query path expressions (`Foo.bar.baz`) are graph traversal operators and, by definition, always return a _distinct set_. 

I would be more explicit here to say "...always return a _distinct set_ of objects reachable by the path". This is so that it's clear what definition we refer to.

> This means that there can be no link of the same name between a pair of objects, and this also means that a set of objects pointed to by a `multi` link is always distinct.

First of all, as written, "there can be no link of the same name" is wrong, I assume you mean "no more than one link of the same name". Second, this doesn't follow from the graph traversal, it's simply one more of our requirements. It's _motivated_ by the traversal results. Basically if you had a multi link where every target object is the same, path traversal rules would evaluate this to a single target. So it's unhelpful and misleading to allow more than one link of the same name between the same pair of objects. It's not impossible, we could have instead used link props to differentiate them and somehow used that in path expressions, but that's not how we did things because we thought it was impractical.

> Properties are different, because their value is scalar, and scalars lack true identity, which makes any proper set mechanics on scalar sets to be expensive due to explicit elimination of duplicates.

"lack true identity" -> "don't have the same notion of identity as objects"
Scalars don't have the same notion of identity as objects, but they don't lack one. They lack an `id` prop specifically. Scalar value itself could be their identity, just not as useful for our purposes.

If I understand correctly, the reason why scalars (typically) require a separate `DISTINCT` operation is because they aren't _joined_ they already are in the same table as yet another column, so we don't get that "for free".

> Furthermore, non-distinct scalar sets are actually _desirable_ in many queries, especially where analytics and tuples are involved.

This is the strongest reason why we chose not to fight scalar duplicates in general.

**Reply by tailhook** [reply]
> This is the strongest reason why we chose not to fight scalar duplicates in general.

This is interesting, because it applies to links as well. I.e. consider linked-in like site:
```
SELECT count(User.skill)
```
If `skill` is a link, actually counts unique skills assigned to user, to count all links, you do something like this:
```
SELECT count((User.id, User.skill))
```

And this equally applies to multi-properties. If we say that multi-properties are automatically distinct, the same query would help.

Even more confusing that:
```
SELECT count(User.some_link.some_prop)
```
Here `some_link` is deduplicated but some_prop isn't.

So my point in this comment: this being different between properties and links only makes it confusing to users for the sake of fixing technical problem. It's not that intuitive and will be hard to memorize for people not participating in this discussion.

**Comment by vpetrovykh** [comment]
> As outlined above, non-distinct `multi` properties are useful in queries, and so from the rule of equivalence of computed and material relations, stored `multi` properties should be allowed to be non-distinct as well.

I like this equivalence formulation very much! I think it's very elegant.

**Reply by 1st1** [reply]
Yeah, good catch. Looks like a typo; I've fixed it.

**Comment by vpetrovykh** [comment]
I'm confused by the sub-title "why link properties are needed at all". The content below it covers multi props vs other options, but there's nothing on _link props_. I assume the title should have been "Digression: why multi properties are needed at all".

**Comment by tailhook** [comment]
Currently we allow:
```
SELECT X { computable_link := {Y, Y}}
```
Which produces non-distinct set via computable link. Will we start requiring `computable_link := (SELECT DISTINCT {Y, Y})` ?

**Reply by 1st1** [reply]
Quoting the proposal:

> A new distinct qualifier is proposed to mark those multi properties that are required to be distinct.

So I assume you'll be able to say

```
distinct computable_link := (SELECT {Y, Y})
```

which is a syntax sugar.

**Reply by elprans** [reply]
No, computable links are always distinct.  The _properties_ aren't.  To answer @tailhook's question, yes, we will start enforcing set distinctness for computable links.

**Reply by 1st1** [reply]
Ah, I missed that Paul was asking his question about links and not properties. But will we allow a `distinct` modifier on computable multi-properties, as in

```
distinct computable_prop := (SELECT {1, 1, 2})
```

?

**Reply by elprans** [reply]
Yes

**Reply by tailhook** [reply]
> But will we allow a distinct modifier on computable multi-properties

I've thought this is compile-time assertion, no? So to make distinct at runtime you should use `SELECT DISTINCT` anyway.

**Reply by elprans** [reply]
Yes, the proposed `distinct` modifier is a compile-time assertion.  To convince the compiler that your expression is, in fact, distinct, you should either use `DISTINCT` or `ASSERT_DISTINCT`

**Comment by 1st1** [comment]
I think this paragraph is the key:

> Properties are different, because their value is scalar, and scalars lack true identity, which makes any proper set mechanics on scalar sets to be expensive due to explicit elimination of duplicates. Furthermore, non-distinct scalar sets are actually desirable in many queries, especially where analytics and tuples are involved.

So I'm +1.

> The -= in-place UPDATE operation is no longer available for multi properties due to comprehension and implementation challenges. Retaining the original equivalence of <prop> -= <arg> to (SELECT <prop> FILTER <prop> != <arg>) is confusing because it is a proper set operation on a multiset. Changing it to mean <prop>.pop(foo) for foo in <arg> is hard to implement efficiently and might arguably also not be universally intuitive.

Can we not implement both `+=` and `-=` on non-distinct multi props? The assymmetry of having just one of them implemented for non-distinct muti props and both of them for distinct multi props (potentially) bothers me.

**Reply by tailhook** [reply]
Asymmetry doesn't bother me. I.e. in a lot of languages strings can be concatenated with `+` but there is no `-` for strings.

**Comment by vpetrovykh** [comment]
This seems like a good place to document some of the thoughts from a conversation about multiplicity inference, since it's needed to enforce multiplicity of computables.

First of all, multiplicity is _very_ brittle. It's very easy to get an expression for which multiplicity cannot be inferred: `UNION` in general kills inference (although some special cases might be inferrable), something as basic as `+` or `++` can kill inference if there are sets on each side (e.g. {1, 3, 5} + {2, 4}), set-returning functions are often a problem, `array_unpack` specifically is a problem, `FOR` is a problem because it's a bunch of unions, even though we probably expect it to have multiplicity 1 in practice in most cases, but that can be tricky to prove. So the bottom line is that even best effort at inferring multiplicity statically is going to result in a lot of "couldn't infer, so assume MANY". This means that we would need to use an escape hatch for these situations - `DISTINCT`.

Let's first examine when multiplicity actually matters. Basically, with this proposal it only matters for defining _computable_ multi links, _computable_ distinct multi properties and for inserting the regular multi links and distinct multi properties. So in these 4 use-cases the user is likely going to have to add a `DISTINCT` (or `assert distinct`, but more on that later, since we don't yet have this feature) for a fair number of cases. This is often around something that may look trivial, like `DISTINCT ({User.email, System.email})` (even if `email` is in fact, non-delegated exclusive property of the common parent of `User` and `System`). So there will be cases where `DISTINCT` is not so much helping or protecting the user as just signaling a limitation of the inference system. `FOR` is particularly bad in this regard. One of the great uses of `FOR` is actually specifying multi links with link properties. Something like:

```
INSERT Foo {
    bar := (
        FOR x IN {(1, 2), (3, 4), (5, 6)}
        UNION (
            SELECT Bar {
                @linkprop := x.0
            } FILTER .val = x.1
        )
    )
}
```
The above is especially problematic for inference because the set for which multiplicity really needs to be inferred is `x.1` (because it's used in the filter) and it's not present in a direct way, but would need to be statically constructed (which is a bad hack anyway, because it doesn't work if the iterator set is not a literal, but, say, a path expression). Basically, the inferrer will give up on this and the user will have to preface every such operation with a `DISTINCT`.

All this means that the code will be typically be littered with `DISTINCT` every time there's a [computable] multi link. And if you have to write something 90% of the time it sort of becomes a reflex, rather than a meaningful construct. "It's just how you do multi link here".

This leads me to the proposal that perhaps we actually want to implicitly inject an `assert distinct` for all _computable_ multi links/properties, while just leave insertions of multi links/properties up to fail in runtime as per normal (much like they'd fail if a constraint was violated). So this assumes some kind of `assert distinct` run-time feature that can be applied to arbitrary expressions. Since we can still use the inferrer, we can turn any `distinct` or `assert distinct` into a NOP if we already inferred that an expression has multiplicity 1.

Pro: this makes the user EdgeQL cleaner. This is merely saving user the hassle of seeing an error message and then *having* to write that same `assert distinct` or `distinct` themselves. There's really no choice for them, because often re-writing the query is not an option.

Con: it is implicit and it is not zero-cost, so at a glance it may be non-obvious to a user why adding a computable multi link slowed things down. To be fair, analyzing queries without any tools is generally not a great way to proceed and our own tools should definitely indicate that an implicit `distinct` happened in this spot.

**Reply by tailhook** [reply]
My intuition about cost is:
1. Runtime-assert-distinct is fine for insertion...
2. But not for querying computable links, because this is kinda "no-opt-out" case, where the best thing you can do is `SELECT DISCTINCT` as that is faster than runtime-assert-distinct.

So as an option, we could allow "intermediate" in-memory non-distinct computables that are collapsed or asserted on insertion. Collapsing especially makes sense because `+=` is union. I.e.:
```
UPDATE X SET { a := A1 };
UPDATE X SET { a += A2 };
UPDATE X SET { a += A3 };
```
I think we allow intersections between `A1` and `A2`, right? At least this is what I'd expect. So this should work the same:
```
UPDATE X SET { a := {A1, A2, A3} };
```
I'm not sure how link properties should work in this case though :(

**Reply by elprans** [reply]
> So as an option, we could allow "intermediate" in-memory non-distinct computables that are collapsed or asserted on insertion. 

This breaks the "equivalence principle", as computed links in aliases would not be guaranteed to be distinct, while regular links are.

The hypothesis here is that computed `multi` links usually have fairly small cardinality, and so runtime multiplicity assertion would be reasonably fast. (Quick testing shows that using a window agg is reasonably efficient).

> Collapsing especially makes sense because `+=` is union

Implicit collapse is problematic, because it may result in data loss.  Consider:

```
INSERT Foo { 
  bar := (
    FOR x in {(A, 'foo'), (A, 'bar')} 
    UNION (
      SELECT Obj {
        @prop := x.1
      } FILTER Obj = x.0
    )
  )
} 
```

If we do silent collapse, as opposed to a "multiplicity violation" exception, then the fact that you lost link property values would be hidden.

**Reply by tailhook** [reply]
> The hypothesis here is that computed multi links usually have fairly small cardinality, and so runtime multiplicity assertion would be reasonably fast. (Quick testing shows that using a window agg is reasonably efficient).

I'm not sure that hypothesis is true, however. I think **stored** links to have fairly small cardinality, but computed may have good use-cases with large sets. For example:
```
type User {
    multi link friends -> User;
    multi link follower -> User;
    multi link friends_of_friends = .friends.friends;
    multi link followers_of_followers = {.friends.friends, .followers.followers}; # friends are kinda followers too
}
```
In typical social network with `max_len(friends) is 5000`, maximum cardinality of `friends_of_friends` is 25'000'000, which is quite huge for sorting. Filtering this set is quite cheap, though, as `friends_of_friends` is a path and we know distinctness of that.

But with `followers_of_followers` this is a much bigger issue as even a simple `EXISTS (.followers_of_followers = User2)` is slow, because it has to make a distinct set first. Am I wrong?

**Comment by vpetrovykh** [comment]
Also, multiplicity inference in the presence of constant extraction is almost useless because something as simple as `SELECT Foo {distinct x := {1, 2}}` will instead become a generic `SELECT Foo {distinct x := {<int64>$param1, <int64>$param2}}` which cannot be proven to be of multiplicity of 1. In order to make this work better, we'd need to teach constant extractor about sets, make it analyze the actual constants, and somehow indicate to the compiler that the "set" parameter is either distinct or not (so that the compiler can use the appropriate query from cache or error out).

**Reply by tailhook** [reply]
Does it have chances to be a common pattern? I would expect:
```
x := array_unpack(<array<int>>$param)
```
to be more common.

**Reply by elprans** [reply]
Yeah, literal sets are not common.  We can improve the situation at some point in the future by allowing `set of` parameters and having the constant extractor track multiplicity factor in constant sets.  Again, this is a fringe case.
